runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineRunner.java

```diff
@@ -67,14 +68,18 @@ public class FlinkPipelineRunner implements PortablePipelineRunner {
   public FlinkPipelineRunner(
       FlinkPipelineOptions pipelineOptions, @Nullable String confDir, List<String> filesToStage) {
     this.pipelineOptions = pipelineOptions;
-    this.confDir = confDir;
+    // confDir takes precedence than pipelineOptions.getFlinkConfDir
```

Comment:
Thanks, done. This is a fix, originally pipelineOptions's flinkConfDir did not take effect at all.

Reply:
Shouldn't pipeline option take precedence though?

Reply:
At a first glance I understood `confDir` parameter as an override to `pipelineOptions.getFlinkConfDir` in `FlinkPipelineRunner's` constructor.

One use case is that the command lines argument of
https://github.com/apache/beam/blob/3edff6a7cf3c57e4589837f2714dbdd981a32063/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineRunner.java#L160
could pass to `FlinkPipelineRunnerConfiguration ` and finally appears as `confDir` parameter here. According to  the comment of https://github.com/apache/beam/blob/3edff6a7cf3c57e4589837f2714dbdd981a32063/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineRunner.java#L213
I understand that if `confDir` is set it should take precedence.

Reply:
Found  origins of this flag: 
From: https://issues.apache.org/jira/browse/BEAM-14492 
```
Sometimes it is necessary to be able to set any flink option via PipelineOptions to the runner - mostly when we submit job from vanilla Java, not being run via flink run
```
https://github.com/apache/beam/pull/17715

I think the intent is consistent with the usage that if a user specifies this via pipeline options, it should take precedence over predetermined runner configuration. 
cc'ing @je-ik just in case and FYI since you mention that ```originally pipelineOptions's flinkConfDir did not take effect at all.```

Reply:
Thanks! That makes sense to me. confDir is per server and pipelineOptions is per pipeline. Will change.

Reply:
Thanks, please update the Javadoc as well.

Reply:
I observe the option does not take effect when invoking a Python pipeline and I targeted the code here. It may be related to What @kennknowles mentioned below there is a "non-portable mode" and also a "portable mode" having different entry points for environment settings.

Reply:
Running Python pipelines on Flink Runner should always be a portable mode.

Reply:
I see, then this is in portable mode confDir settings in pipeline options did not take effect.